physics/hierarchical_system_body.hpp

```diff
@@ -0,0 +1,151 @@
+#pragma once
+
+#include "physics/hierarchical_system.hpp"
+
+#include <iterator>
+
+namespace principia {
+namespace physics {
+
+template<typename Frame>
+HierarchicalSystem<Frame>::HierarchicalSystem(
+    not_null<std::unique_ptr<MassiveBody const>> primary)
+    : system_(std::move(primary)) {
+  systems_[system_.primary.get()] = &system_;
+}
+
+template<typename Frame>
+void HierarchicalSystem<Frame>::Add(
+    not_null<std::unique_ptr<MassiveBody const>> body,
+    not_null<MassiveBody const*> const parent,
+    KeplerianElements<Frame> const& jacobi_osculating_elements) {
+  not_null<MassiveBody const*> unowned_body = body.get();
+  System& parent_system = *systems_[parent];
+  parent_system.satellites.emplace_back(
+      make_not_null_unique<Subsystem>(std::move(body)));
+  {  // Hide the moved-from |body|.
+    not_null<MassiveBody const*> body = unowned_body;
+    not_null<Subsystem*> inserted_system =
+        parent_system.satellites.back().get();
+    systems_[body] = inserted_system;
+    inserted_system->jacobi_osculating_elements = jacobi_osculating_elements;
+  }
+}
+
+template<typename Frame>
+typename HierarchicalSystem<Frame>::BarycentricSystem
+HierarchicalSystem<Frame>::Get() {
+  auto const semimajor_axis_less_than = [](
+      not_null<std::unique_ptr<Subsystem>> const& left,
+      not_null<std::unique_ptr<Subsystem>> const& right) -> bool {
+    return left->jacobi_osculating_elements.semimajor_axis <
+           right->jacobi_osculating_elements.semimajor_axis;
+  };
+
+  // Data about a |Subsystem|.
+  struct BarycentricSubystem {
```

Comment:
I'd rather declare this type in the private part of the class.  Even though it's only used here, it would make it easier to see the parallel with BarycentricSystem.